lib/src/merged_tree.rs

```diff
@@ -971,6 +974,235 @@ impl Iterator for TreeDiffIterator<'_> {
+        let store = tree1.store().clone();
+        let mut stream = Self {
+            store,
+            matcher,
+            legacy_format_before: matches!(tree1, MergedTree::Legacy(_)),
+            legacy_format_after: matches!(tree2, MergedTree::Legacy(_)),
+            items: BTreeMap::new(),
+            pending_trees: VecDeque::new(),
+            // TODO: maybe the backends can suggest the conurrency limit?
+            max_concurrent_reads: 100,
+            max_queued_items: 10000,
+        };
+        stream.add_dir_diff_items(RepoPath::root(), Ok((tree1, tree2)));
+        stream
+    }
+
+    /// Gets the given tree if `value` is a tree, otherwise an empty tree.
+    async fn tree(
+        store: Arc<Store>,
+        legacy_format: bool,
+        dir: RepoPath,
+        values: MergedTreeValue,
+    ) -> BackendResult<MergedTree> {
+        let trees = if values.is_tree() {
+            let builder: MergeBuilder<Tree> = futures::stream::iter(values.iter())
+                .then(|value| TreeDiffIterator::single_tree(&store, &dir, value.as_ref()))
+                .try_collect()
+                .await?;
+            builder.build()
+        } else {
+            Merge::resolved(Tree::null(store.clone(), dir.clone()))
+        };
+        // Maintain the type of tree, so we resolve `TreeValue::Conflict` as necessary
+        // in the subtree
+        if legacy_format {
+            Ok(MergedTree::Legacy(trees.into_resolved().unwrap()))
+        } else {
+            Ok(MergedTree::Merge(trees))
+        }
+    }
+
+    fn add_dir_diff_items(
+        &mut self,
+        dir: RepoPath,
+        tree_diff: BackendResult<(MergedTree, MergedTree)>,
+    ) {
+        let (tree1, tree2) = match tree_diff {
+            Ok(trees) => trees,
+            Err(err) => {
+                self.items.insert(DiffStreamKey::normal(dir), Err(err));
+                return;
+            }
+        };
+
+        for basename in merged_tree_basenames(&tree1, &tree2) {
+            let value_before = tree1.value(basename);
+            let value_after = tree2.value(basename);
+            if value_after != value_before {
+                let path = dir.join(basename);
+                let before = value_before.to_merge();
+                let after = value_after.to_merge();
+                if (before.is_tree() || after.is_tree()) && !self.matcher.visit(&path).is_nothing()
+                {
+                    let before_tree_future = Self::tree(
+                        self.store.clone(),
+                        self.legacy_format_before,
+                        path.clone(),
+                        before.clone(),
+                    );
+                    let after_tree_future = Self::tree(
+                        self.store.clone(),
+                        self.legacy_format_after,
+                        path.clone(),
+                        after.clone(),
+                    );
```

Comment:
We store the returned future in `self.pending_trees`, so that doesn't seem easy if it captures references to `tree1`/`tree2`. Maybe there's a way that I'm missing?

Reply:
I just mean the `store` can be cloned from `tree1`/`tree2`, and `self.store` can be removed. (If `Self::tree()` is a method, it makes sense to cache the tree to remove the `tree` argument, but it's not.)

Reply:
Done.